src/index.ts

```diff
@@ -0,0 +1,87 @@
+export type StackErrorJSON = {
+  code: string
+  message: string
+  metadata: Record<string, unknown>
+  name: string
+  stack: Array<StackErrorJSON>
+}
+
+export class StackError extends Error {
+  static fromJSON(json: StackErrorJSON): StackError {
+    const error = new StackError(json.code, json.message)
+    error.errorStack = (json.stack ?? []).reduceRight((stack, e) => {
+      const err = StackError.fromJSON(e)
+      err.errorStack = stack
+      return [err, ...stack]
+    }, [] as Array<StackError>)
+    error.metadata = json.metadata ?? {}
+    error.name = json.name ?? 'StackError'
+    return error
+  }
+
+  code: string
+  errorStack: Array<StackError>
+  metadata: Record<string, unknown> = {}
+  name = 'StackError'
+
+  constructor(code: string, message: string, parentError?: StackError) {
+    super(message)
+    Object.setPrototypeOf(this, StackError.prototype)
+    if (Error.captureStackTrace) {
+      Error.captureStackTrace(this, StackError)
+    }
+
+    this.code = code
+    this.errorStack = parentError ? parentError.toErrorStack() : []
+  }
+
+  toErrorStack(): Array<StackError> {
+    return [this, ...this.errorStack]
+  }
+
+  toJSON(withStack = true): StackErrorJSON {
+    return {
+      code: this.code,
+      message: this.message,
+      metadata: this.metadata,
+      name: this.name,
+      stack: withStack ? this.errorStack.map(e => e.toJSON(false)) : []
+    }
+  }
+
+  toString(): string {
+    return `[${this.code}] ${this.message}`
+  }
+}
+
+// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
+export function createNamespaceError(namespace: string, metadata: Record<string, unknown> = {}) {
+  return class NamespaceError extends StackError {
+    constructor(code: string | number, message: string, parentError?: StackError) {
+      super(`${namespace}${code}`, message, parentError)
+      Object.setPrototypeOf(this, NamespaceError.prototype)
+      if (Error.captureStackTrace) {
+        Error.captureStackTrace(this, NamespaceError)
+      }
+
+      this.metadata = metadata
+    }
+  }
+}
+
+export function assert(condition: boolean, message = 'Assertion failed'): asserts condition {
+  if (!condition) {
+    throw new Error(message)
```

Comment:
I think it might be nice if we had a common type that was the base class of all of our custom ErrorClasses, that way we can distinguish Errors generated by our code those thrown from libraries we are using.  Something like `CeramicError` that extends Error and all of our specific error classes extend in turn

Reply:
Maybe that's actually what StackError is?  Maybe we rename StackError to CeramicError?  Or we could have CeramicError extend StackError

Reply:
I imagine it's a library we could use in any of our libraries like the DID and IDX ones, maybe other utilities etc. so I picked a name more generic than `CeramicError` but we can rename it as needed if you have any other suggestion?

Reply:
Okay I see now, I think I was thinking of this all at one layer off from how it's actually designed to be used.  You're suggesting that there's only one call to `createNamespaceError` per package, right?  And then all the errors thrown within that package either use that error or subclasses of that error.  I think that makes sense.  Maybe this should be `createErrorForPackage` instead of `createNamespaceError`?

Reply:
I don't think there's necessarily a 1:1 mapping between error namespaces and packages, see my example above with `ProtocolError` and `InternalError` serving different use cases, or on the other side I imagine I could use a single namespace for IDX across the different packages.
I don't mind changing the name if you think something else is more explicit, but I'm not sure `package` is necessarily the right abstraction here.

Reply:
I guess what I'm trying to figure out is what is a new Error type and what is a different error code within the same Error type?

Reply:
That's up to us to decide this for each project I think.
We could use a single namespace/error type for all Ceramic packages, one per package or even multiple per package (as in the example above) depending on preferences, this library doesn't enforce any specific pattern.